### PR TITLE
Global variables

### PR DESCRIPTION
**Pull request by pfitzseb**
image

This simply adds the current function's defining module as a scope.

### CODE REVIEW

File: src/debugger_requests.jl [hunk 1]
```diff
@@ -352,11 +352,20 @@ function scopes_request(conn, state::DebuggerState, params::ScopesArguments)
...
+
     if isfile(file_name) && code_range !== nothing
-        return ScopesResponseArguments([Scope(name = "Local", variablesReference = var_ref_id, expensive = false, source = Source(name = basename(file_name), path = file_name), line = code_range.start, endLine = code_range.stop)])
+        push!(scopes, Scope(name = "Local", variablesReference = var_ref_id, expensive = false, source = Source(name = basename(file_name), path = file_name), line = code_range.start, endLine = code_range.stop))
     else
-        return ScopesResponseArguments([Scope(name = "Local", variablesReference = var_ref_id, expensive = false)])
+        push!(scopes, Scope(name = "Local", variablesReference = var_ref_id, expensive = false))
     end
+
+    curr_mod = JuliaInterpreter.moduleof(curr_fr)
```
Comment (aviatesk) [inline src/debugger_requests.jl:363]: I forgot the detail, but does `moduleof` works for toplevel scope ?
Or we don't support that yet ?

Comment (pfitzseb) [reply src/debugger_requests.jl:363]: It does -- `moduleof ` is a no-op when given a module.

Comment (aviatesk) [reply src/debugger_requests.jl:363]: aha, yes:

https://github.com/JuliaDebug/JuliaInterpreter.jl/blob/master/src/utils.jl#L6

**Comment by davidanthoff** [comment]
Hm, I would have thought that for a given stackframe we would show the global variables that are used in that function, i.e. either read or written in the code for that stackframe? If I have a function like
```julia
function foo()
  println(Main.Foo.Bar.varY)
  global varX = "Hello"
end
```
I would have thought we would show `varY` and `varX` for that stackframe as the globals. That seems more useful than the global variables of the module where the function is defined, because if they aren't accessed by the function they seem pretty irrelevant?

For the "global frame" we could show all the globals, so it would essentially end up as the equivalent of the workspace stuff we have for the REPL? It would also make it pretty easy to access any global variable via the global frame, right?

**Comment by pfitzseb** [comment]
Fair enough. I've implemented a somewhat limited version of this (doesn't pick up on `const` globals and assignments that aren't explicitly declared `global`):
image

The `Global ($mod)` section can always be useful, I think -- it allows you to inspect globals in modules you've stepped into, e.g. `Random` in the first screenshot posted here (which isn't necessarily reachable from the global frame).

**Comment by pfitzseb** [comment]
Alternative:
image

Let me know which one you like better!

**Comment by davidanthoff** [comment]
I definitely like the version that does not show the functions one is calling better, i.e. the second one.

The way I had thought we would do this is that for a given stack frame, we have one scope "Local", and one scope "Global". The "Local" scope would not contain any global variables, but only the local ones (and we could eventually consider splitting out "Arguments" and "closures" as yet more scopes). The "Global" scope would have the global vars that are used in the current stack frame. I think that is how other debuggers do it? But I also hear you about the usefulness of a global scope that surrounds the function. So maybe we should have the following scopes: `Local`, `Global`, `Global ($ParentModule)`?

In any case, I think we should merge this and then we can iterate on these things in future PRs, this is such a major improvements, regardless how it is implemented in the details, that we shouldn't hold it up :) I'll approve now and then you can merge once you have sorted out the difference between the two versions you proposed.

**Comment by pfitzseb** [comment]
Ok, this is the current state:
image

It relies on the (somewhat imprecise) heuristic that the user probably isn't interested in globals that are being called (i.e. functions).

**Comment by davidanthoff** [comment]
Perfect!

**Pull request merged**